src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+
+            if (serviceLifetime == ServiceLifetime.Transient && typeof(IDisposable).IsAssignableFrom(typeof(TSchema)))
+            {
+                // This scenario can cause a memory leak if the schema is requested from the root service provider.
+                // If it was requested from a scoped provider, then there is no reason to register it as transient.
+                // See following link:
+                // https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#disposable-transient-services-captured-by-container
+                throw new InvalidOperationException("A schema that implements IDisposable cannot be registered as a transient service.");
+            }
+
+            // Register the service with the DI provider as TSchema, overwriting any existing registration
+            builder.Register(schemaFactory, serviceLifetime);
+
+            // Now register the service as ISchema if not already registered.
+            builder.TryRegister<ISchema>(schemaFactory, serviceLifetime);
+
+            return builder;
+        }
+        #endregion
+
+        #region - AddDocumentExecuter -
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter, TDocumentExecuter>(ServiceLifetime.Singleton);
+
+        /// <summary>
+        /// Registers <paramref name="documentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder, TDocumentExecuter documentExecuter)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter>(documentExecuter ?? throw new ArgumentNullException(nameof(documentExecuter)));
+
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework. The supplied factory method is used to create the document executer.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder, Func<IServiceProvider, TDocumentExecuter> documentExecuterFactory)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter>(documentExecuterFactory ?? throw new ArgumentNullException(nameof(documentExecuterFactory)), ServiceLifetime.Singleton);
+        #endregion
+
+        #region - AddComplexityAnalyzer -
```

Comment:
I would split that big class into small ones  - `DocumentExecuterGraphQLBuilderExtensions`, `ComplexityAnalyzerGraphQLBuilderExtensions` and so on.

Reply:
After reading the entire file, I **highly recommend** splitting methods.

Reply:
We could split the file and use partial classes for v4.  In v5, we can consider multiple classes.

Reply:
I would split only for v5 and keep as is for v4 to eliminate double work.

Reply:
#2790